README.md

<a href="https://www.npmjs.com/package/watch-state">
  <img src="https://raw.githubusercontent.com/d8corp/watch-state/v3.3.3/img/logo.svg" align="left" width="90" height="90" alt="Watch-State logo by Mikhail Lysikov">
</a>

# &nbsp; @watch-state/fetch

&nbsp;

[![NPM](https://img.shields.io/npm/v/@watch-state/fetch.svg)](https://www.npmjs.com/package/@watch-state/fetch)
[![minzipped size](https://img.shields.io/bundlephobia/minzip/@watch-state/fetch)](https://bundlephobia.com/result?p=@watch-state/fetch)
[![downloads](https://img.shields.io/npm/dm/@watch-state/fetch.svg)](https://www.npmtrends.com/@watch-state/fetch)
[![changelog](https://img.shields.io/badge/changelog-⋮-brightgreen)](https://changelogs.xyz/@watch-state/fetch)
[![license](https://img.shields.io/npm/l/@watch-state/fetch)](https://github.com/d8corp/watch-state-fetch/blob/main/LICENSE)

Fetch with [watch-state](https://www.npmjs.com/package/watch-state)  
Based on [@watch-state/async](https://www.npmjs.com/package/@watch-state/async)

[![stars](https://img.shields.io/github/stars/d8corp/watch-state-fetch?style=social)](https://github.com/d8corp/watch-state-fetch/stargazers)
[![watchers](https://img.shields.io/github/watchers/d8corp/watch-state-fetch?style=social)](https://github.com/d8corp/watch-state-fetch/watchers)

### Install

npm
```bash
npm i @watch-state/fetch
```

yarn
```bash
yarn add @watch-state/fetch
```

### Usage

`Fetch` is a `Promise` like constructor
```javascript
import Fetch from '@watch-state/fetch'

const user = await new Fetch('https://reqres.in/api/users/1')
```

### then, catch, finally

Use `then`, `catch` and `finally` like with `Promise`

```javascript
const user = new Fetch('https://reqres.in/api/users/1')
user
  .finally(value => console.log('finally', value))
  .then(value => console.log('then', value))
  .catch(value => console.log('catch', value))
```

### loading

You may check status of the request with `loading` field,
it's `true` when data is loading. This field is observable.

```javascript
const user = new Fetch('https://reqres.in/api/users/1')
// user.loading === true

await user
// user.loading === false
```

### loaded

You may check status of the request with `loaded` field,
it's `true` if the data is loaded at least one time.
This is an observable field.

```javascript
const user = new Fetch('https://reqres.in/api/users/1')
// user.loaded === false

await user
// user.loaded === true
```

### value

You may get result with `value`.
This is an observable field.

```javascript
const user = new Fetch('https://reqres.in/api/users/1')

new Watch(() => console.log(user.loading ? 'loading...' : user.value))
// Watch from watch-state
```

### error

You may handle errors or responses with error status by `error` field.  
This is an observable field.

```javascript
const user = new Fetch('https://reqres.in/api/users/23')
// returns 404 user not found

new Watch(() => console.log(
  user.loading ? 'loading...' : user.error || user.value
))
```

### type

You can convert the response to `json` | `text` | `blob` | `arrayBuffer` | `formData`.

```javascript
const user = new Fetch('https://reqres.in/api/users/1', {
  type: 'text'
})
```

### default

You may provide default `value` for `Fetch`

```javascript
const user = new Fetch('https://reqres.in/api/users/1', {
  default: {
    data: { id: null }
  }
})

// user.value.data.id === null

await user
// user.value.data.id === 1
```

### response

`response` is the same `value` but without default value.  
This is an observable field.

```javascript
const user = new Fetch('https://reqres.in/api/users/1', {
  default: {
    data: { id: null }
  }
})

// user.value.data.id === null
// user.response === undefined

await user
// user.value.data.id === 1
// user.response.data.id === 1
```

### update

Unlike a `Promise`, you may reuse `Fetch` with `update` method

```javascript
const user = new Fetch('https://reqres.in/api/users/1')

await user
// request to https://reqres.in/api/users/1

user.update()

await user
// request to https://reqres.in/api/users/1
```

### resolve

You may use `resolve` to set value without updating or loading.

```javascript
const user = new Fetch('https://reqres.in/api/users/1')

user.resolve({ data: { id: 2 } })

await user
// user.value.data.id === 2

user.update()

await user
// user.value.data.id === 1
```

### reject

The same as resolve, but it set an `error` instead of `value`.

### on, once, off

You may add a listener to react on the next events `resolve` | `reject` | `update`

```javascript
const user = new Fetch('https://reqres.in/api/users/1')
let test = false

user.on('resolve', value => test = value)

await user
// test === '{...}'
```

You may add a listener which reacts only once with `once` method.

```javascript
const user = new Fetch('https://reqres.in/api/users/1')
let test = false

user.once('resolve', value => test = value)

await user
// test === '{...}'
```

You may turn off a listener

```javascript
const user = new Fetch('https://reqres.in/api/users/1')
let test = false

const listenr = e => test = e

user.once('resolve', listenr)
user.off('resolve', listenr)

await user
// test === undefined
```

### answer

If you want to know the status of the response or headers you can use `answer` field.

```javascript
const user = new Fetch('https://reqres.in/api/users/23')

// user.answer === undefined

await user

// user.answer.ok === false
// user.answer.status === 404
```

It's better if you extend `Fetch` class to get it.
```typescript
class MyFetch extends Fetch {
  @cache get status (): number {
    return this.answer?.status
  }
}

const user = new MyFetch('https://reqres.in/api/users/23')

// user.status === undefined

await user

// user.status === 404
```

## Issues

If you find a bug, please file an issue on [GitHub](https://github.com/d8corp/watch-state-fetch/issues)  

[![issues](https://img.shields.io/github/issues-raw/d8corp/watch-state-fetch)](https://github.com/d8corp/watch-state-fetch/issues)


CHANGELOG.md
# Changelog

## v1.0

### v1.0.3 [![08.01.2022](https://img.shields.io/date/1641642518)](https://github.com/d8corp/watch-state-ajax/tree/v1.0.3)
- update dependencies
- added `update` argument of `Fetch` constructor
- prevent body parsing with `204` status

### v1.0.2 [![07.01.2022](https://img.shields.io/date/1641559348)](https://github.com/d8corp/watch-state-ajax/tree/v1.0.2)
- update dependencies

### v1.0.1 [![03.01.2022](https://img.shields.io/date/1641233980)](https://github.com/d8corp/watch-state-ajax/tree/v1.0.1)
- improved types

### v1.0.0 [![03.01.2022](https://img.shields.io/date/1641232894)](https://github.com/d8corp/watch-state-ajax/tree/v1.0.0)
- added base functionality


src/index.test.ts
import Fetch from '.'
import 'isomorphic-fetch'

const user1 = fetch('https://reqres.in/api/users/1').then(data => data.json())

describe('Ajax', () => {
  test('simple', async () => {
    expect(await new Fetch('https://reqres.in/api/users/1')).toEqual(await user1)
  })
  test('loading', async () => {
    const ajax = new Fetch('https://reqres.in/api/users/1')
    expect(ajax.loading).toBe(true)
    await ajax
    expect(ajax.loading).toBe(false)
  })
  test('loaded', async () => {
    const ajax = new Fetch('https://reqres.in/api/users/1')
    expect(ajax.loaded).toBe(false)
    await ajax
    expect(ajax.loaded).toBe(true)
  })
  test('value', async () => {
    const ajax = new Fetch('https://reqres.in/api/users/1')
    expect(ajax.value).toBe(undefined)
    await ajax
    expect(ajax.value).toEqual(await user1)
  })
  test('then catch finally', async () => {
    const user = new Fetch('https://reqres.in/api/users/1')
    let f, t, c

    user
      .finally(value => f = value)
      .then(value => t = value)
      .catch(value => c = value)

    await user

    expect(f).toEqual(await user1)
    expect(t).toEqual(await user1)
    expect(c).toBe(undefined)
  })
  test('error', async () => {
    const user = new Fetch('https://reqres.in/api/login', {method: 'post', type: "json"})
    try { await user } catch (e) {}
    expect(user.value).toBe(undefined)
    expect(user.error).toEqual({error: 'Missing email or username'})
  })
  test('answer', async () => {
    const user = new Fetch('https://reqres.in/api/users/23')
    expect(user.answer).toBe(undefined)
    try { await user } catch (e) {}
    expect(user.answer.ok).toBe(false)
    expect(user.answer.status).toBe(404)
  })
  test('resolve', async () => {
    const user = new Fetch('https://reqres.in/api/users/1')
    user.resolve({data: {id: 2}})
    await user
    expect(user.value.data.id).toBe(2)

    user.update()

    await user

    expect(user.value.data.id).toBe(1)
  })
})


lib/index.js
'use strict';

var tslib = require('tslib');
var Async = require('@watch-state/async');
var watchState = require('watch-state');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var Async__default = /*#__PURE__*/_interopDefaultLegacy(Async);

var Fetch = /** @class */ (function (_super) {
    tslib.__extends(Fetch, _super);
    function Fetch(url, options, update) {
        if (options === void 0) { options = {}; }
        var _this = _super.call(this, tslib.__assign(tslib.__assign({}, options), { request: function (resolve, reject) {
                var _a = options.type, type = _a === void 0 ? 'json' : _a;
                var answer;
                fetch(url, options).then(function (data) {
                    answer = data;
                    return data.status === 204 ? undefined : data[type]();
                }).then(function (data) {
                    watchState.globalEvent.start();
                    _this.answer = answer;
                    answer.status > 399 ? reject(data) : resolve(data);
                    watchState.globalEvent.end();
                }, function (e) {
                    watchState.globalEvent.start();
                    _this.answer = answer;
                    reject(e);
                    watchState.globalEvent.end();
                });
            } }), update) || this;
        _this._answer = new watchState.State();
        return _this;
    }
    Object.defineProperty(Fetch.prototype, "answer", {
        get: function () {
            return this._answer.value;
        },
        set: function (value) {
            this._answer.value = value;
        },
        enumerable: false,
        configurable: true
    });
    return Fetch;
}(Async__default["default"]));

module.exports = Fetch;


lib/index.es6.js
import Async from '@watch-state/async';
import { globalEvent, State } from 'watch-state';

class Fetch extends Async {
    constructor(url, options = {}, update) {
        super(Object.assign(Object.assign({}, options), { request: (resolve, reject) => {
                const { type = 'json' } = options;
                let answer;
                fetch(url, options).then(data => {
                    answer = data;
                    return data.status === 204 ? undefined : data[type]();
                }).then(data => {
                    globalEvent.start();
                    this.answer = answer;
                    answer.status > 399 ? reject(data) : resolve(data);
                    globalEvent.end();
                }, e => {
                    globalEvent.start();
                    this.answer = answer;
                    reject(e);
                    globalEvent.end();
                });
            } }), update);
        this._answer = new State();
    }
    get answer() {
        return this._answer.value;
    }
    set answer(value) {
        this._answer.value = value;
    }
}

export { Fetch as default };


src/index.ts
import Async, { IAsyncOptions } from '@watch-state/async'
import { State, globalEvent } from 'watch-state'

export interface FetchOptions<V = any, E = any> extends RequestInit, Omit<IAsyncOptions<V, E>, 'request'> {
  type?: 'json' | 'text' | 'blob' | 'arrayBuffer' | 'formData'
}

export default class Fetch<V = any, E = any> extends Async<V, E> {
  _answer = new State()
  get answer () {
    return this._answer.value
  }
  set answer (value) {
    this._answer.value = value
  }

  constructor (url: string, options: FetchOptions<V, E> = {}, update?: boolean) {
    super({...options, request: (resolve, reject) => {
      const { type = 'json' } = options
      let answer
      fetch(url, options).then(data => {
        answer = data
        return data.status === 204 ? undefined : data[type]()
      }).then(data => {
        globalEvent.start()
        this.answer = answer
        answer.status > 399 ? reject(data) : resolve(data)
        globalEvent.end()
      }, e => {
        globalEvent.start()
        this.answer = answer
        reject(e)
        globalEvent.end()
      })
    }}, update)
  }
}
